src/runtime_src/core/pcie/linux/system_linux.cpp

```diff
@@ -36,22 +38,32 @@
 
 namespace {
 
+xrt_core::system_linux* singleton = nullptr;
+
 // Singleton registers with base class xrt_core::system
-// during static global initialization.  If statically
+// during static global initialization. If statically
 // linking with libxrt_core, then explicit initialiation
 // is required
-static xrt_core::system_linux*
+static inline xrt_core::system_linux*
 singleton_instance()
 {
-  static xrt_core::system_linux singleton;
-  return &singleton;
+  if (!singleton)
+    static xrt_core::system_linux s;
+
+  if (singleton)
+    return singleton;
+
+  throw std::runtime_error("system_linux singleton is not initialized");
 }
 
 // Dynamic linking automatically constructs the singleton
+// Do not instantiate singleton, if SHIM is extended from outside.
+#ifndef EXTERNAL_SHIM
```

Comment:
didn't you remove EXTERNAL_SHIM?

Reply:
This is still needed until we can really support driver plug-in .so libs. Today, the system_linux class instance implemented here only knows about pci_driver_xocl and pci_driver_xclmgmt. A derived system class instance knows new drivers and should register itself as the singleton. This code is in the extended part in external repo, where EXTERNAL_SHIM will be defined.